### PR TITLE
NLP implementation client update

### PR DESCRIPTION
**Pull request by KamenDimitrov97**
### What
I've added NLP implementation to dp-search-api with a feature toggle 
I've removed NLP clients from dp-search-api and added them into dp-api-clients-go.
Started using the clients from dp-api-client-go.
Fixed previous tests and added new ones 

### How to review

- [ ]  Run tests
- [ ]  Consistency
- [ ]  Typos
- [ ]  Comments
- [ ]  Syntax order

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -37,6 +37,15 @@ environment variables, or with a link to a configuration guide.
 | AWS_TLS_INSECURE_SKIP_VERIFY | false                      | This should never be set to true, as it disables SSL certificate verification. Used only for development           
 | BIND_ADDR                    | :23900                     | The host and port to bind to                                                                                       
 | ELASTIC_SEARCH_URL           | "<http://localhost:11200>" | Http url of the ElasticSearch server                                                                               
+| -------------------          | -------------------       | -----------------------------------
+| NLP_BERLIN_API_ENDPOINT      | "/v1/berlin/search"       | Http url of the Berlin server
+| NLP_BERLIN_API_URL	         | "http://localhost:28900"  | Http url of the Berlin server
```
Comment (lindenmckenzie) [inline README.md:42]: whitespace issues

File: README.md [hunk 2]
```diff
@@ -37,6 +37,15 @@ environment variables, or with a link to a configuration guide.
...
+| NLP_CATEGORY_API_ENDPOINT    | "/categories"             | Http url of the Category server
+| NLP_CATEGORY_API_URL	       | "http://localhost:28800"  | Http url of the Category server
+| NlpToggle           	       | "true"                    | Toggles NLP querying on/off
+| NLP_SCRUBBER_API_ENDPOINT    | "/v1/scrubber"            | Http url of the Scrubber server
+| NLP_SCRUBBER_API_URL	       | "http://localhost:28700"  | Http url of the Scrubber server
```
Comment (lindenmckenzie) [inline README.md:47]: These do not match config.go

File: api/api.go [hunk 3]
```diff
@@ -116,9 +138,11 @@ func (a *SearchAPI) RegisterGetSearchReleases(validator QueryParamValidator, bui
...
 			builder,
-			a.dpESClient,
+			a.clList.dpESClient,
 			transformer,
 		),
 	).Methods(http.MethodGet)
 	return a
 }
+
+// as well as the API's elasticsearch client and respo
```
Comment (lindenmckenzie) [inline api/api.go:148]: what's this comment doing here?

File: api/api.go [hunk 4]
```diff
@@ -69,26 +83,34 @@ type ReleaseResponseTransformer interface {
...
+		clList:      clientList,
+		permissions: permissions,
 	}
 }
 
 // RegisterGetSearch registers the handler for GET /search endpoint
-// with the provided validator and query builder
-// as well as the API's elasticsearch client and response transformer
-func (a *SearchAPI) RegisterGetSearch(validator QueryParamValidator, builder QueryBuilder, transformer ResponseTransformer) *SearchAPI {
+// with the provided validator and query buildernse transformer
```
Comment (lindenmckenzie) [inline api/api.go:105]: malformed comment.

File: api/search.go [hunk 5]
```diff
@@ -155,24 +155,58 @@ func CreateRequests(w http.ResponseWriter, req *http.Request, validator QueryPar
 	contentTypes := defaultContentTypes
 	if contentTypesParam != "" {
 		contentTypes = strings.Split(contentTypesParam, ",")
-		disallowed, err := validateContentTypes(contentTypes)
-		if err != nil {
-			log.Warn(ctx, err.Error(), log.Data{"param": ParamContentType, "value": contentTypesParam, "disallowed": disallowed})
+		disallowed, er := validateContentTypes(contentTypes)
+		if er != nil {
+			log.Warn(ctx, er.Error(), log.Data{"param": ParamContentType, "value": contentTypesParam, "disallowed": disallowed})
 			http.Error(w, fmt.Sprint("Invalid content_type(s): ", strings.Join(disallowed, ",")), http.StatusBadRequest)
```
Comment (lindenmckenzie) [inline api/search.go:161]: why `er` not `err`?

File: api/search.go [hunk 6]
```diff
@@ -155,24 +155,58 @@ func CreateRequests(w http.ResponseWriter, req *http.Request, validator QueryPar
...
+		log.Warn(ctx, err.Error(), log.Data{"param": "fromDate", "value": fromDateParam})
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return "", nil, nil
+	}
+
+	toDateParam := paramGet(params, "toDate", "")
+	toDate, err := validator.Validate(ctx, "date", toDateParam)
+	if err != nil {
+		log.Warn(ctx, err.Error(), log.Data{"param": "toDate", "value": toDateParam})
+		http.Error(w, "Invalid dateTo parameter", http.StatusBadRequest)
```
Comment (lindenmckenzie) [inline api/search.go:186]: param name should match.

File: api/search.go [hunk 7]
```diff
@@ -155,24 +155,58 @@ func CreateRequests(w http.ResponseWriter, req *http.Request, validator QueryPar
...
+		log.Warn(ctx, err.Error(), log.Data{"param": ParamSort, "value": sortParam})
+		http.Error(w, "Invalid sort parameter", http.StatusBadRequest)
+		return "", nil, nil
+	}
+
+	fromDateParam := paramGet(params, "fromDate", "")
+	fromDate, err := validator.Validate(ctx, "date", fromDateParam)
+	if err != nil {
+		log.Warn(ctx, err.Error(), log.Data{"param": "fromDate", "value": fromDateParam})
+		http.Error(w, err.Error(), http.StatusBadRequest)
```
Comment (lindenmckenzie) [inline api/search.go:178]: should give string description to user, not err.Error()

File: config/config.go [hunk 8]
```diff
@@ -37,10 +42,15 @@ func Get() (*Config, error) {
...
 	cfg = &Config{
 		BindAddr:                   ":23900",
+		BerlinAPIURL:               "http://localhost:28900",
+		CategoryAPIURL:             "http://localhost:28800",
 		ElasticSearchAPIURL:        "http://localhost:11200",
 		GracefulShutdownTimeout:    5 * time.Second,
 		HealthCheckCriticalTimeout: 90 * time.Second,
 		HealthCheckInterval:        30 * time.Second,
+		NlpHubSettings:             "{\"categoryWeighting\": 1000000000000000000000.0, \"categoryLimit\": 100, \"defaultState\": \"gb\"}",
+		NlpToggle:                  false,
```
Comment (lindenmckenzie) [inline config/config.go:52]: Please add tests.

File: go.mod [hunk 9]
```diff
@@ -11,73 +11,77 @@ replace google.golang.org/grpc => google.golang.org/grpc v1.55.0
 //     - sonatype-2019-0890 # pkg:golang/github.com/pkg/sftp@v1.10.1
 replace github.com/spf13/cobra => github.com/spf13/cobra v1.7.0
 
+replace github.com/ONSdigital/dp-api-clients-go/v2 => ./../dp-api-clients-go
+
+replace github.com/ONSdigital/dp-search-scrubber-api => ./../onyx-dp-search-scrubber-api
```
Comment (lindenmckenzie) [inline go.mod:16]: will need removing pre-merge.

File: query/templates/search/contentFilterOnURIPrefix.tmpl [hunk 10]
```diff
@@ -1,7 +1,7 @@
 
       {{if .URIPrefix}}
           { "prefix" :
-                { "uri" : "{{.URIPrefix}}" }
+                { "uri" : { "value": "{{.URIPrefix}}", "boost": 10.0 } }
           }
         {{if or .Topic .TopicWildcard}},{{end}}
```
Comment (lindenmckenzie) [inline query/templates/search/contentFilterOnURIPrefix.tmpl:6]: Why this change?

File: query/templates/search/contentQuery.tmpl [hunk 11]
```diff
@@ -1,11 +1,16 @@
 {
+explain: true,
```
Comment (lindenmckenzie) [inline query/templates/search/contentQuery.tmpl:2]: should this be in quote marks? also indentation.

File: query/templates/search/contentQuery.tmpl [hunk 12]
```diff
@@ -1,11 +1,16 @@
...
  {{- if .From -}}
  	"from" : {{- .From}},
  {{- end}}
  "size" : {{.Size}},
  "query" : {
      "bool" : {
+     "should": [
+             {{template "nlpCategory.tmpl" .}}
+        ],
+
```
Comment (lindenmckenzie) [inline query/templates/search/contentQuery.tmpl:13]: why whitespace?

File: query/templates/search/search.tmpl [hunk 13]
```diff
@@ -1,5 +1,3 @@
 {{/* $$ indicates newline, each header and query MUST be on its own line - with a blank line at the end */}}
 {{- template "contentHeader.tmpl" .}}$$
 {{- template "contentQuery.tmpl" .}}$$
-{{- template "countHeader.tmpl" .}}$$
-{{- template "countQuery.tmpl" .}}$$
```
Comment (lindenmckenzie) [inline query/templates/search/search.tmpl:5]: Why this change?

File: query/templates/search/v710/contentQuery.tmpl [hunk 14]
```diff
@@ -6,15 +6,18 @@
...
+			],
+			"must" : {
+				{{- if .Term}}
+					{{- template "weightedQuery.tmpl" .}}
+				{{- else}}
+					{{- template "matchAll.tmpl" .}}
+				{{- end}}
+			}
+			{{template "contentFilters.tmpl" .}}
+		}
```
Comment (lindenmckenzie) [inline query/templates/search/v710/contentQuery.tmpl:20]: Indentation.

File: swagger.yaml [hunk 15]
```diff
@@ -59,6 +59,28 @@ paths:
...
+      produces:
+        - application/json
+      responses:
+        200:
+          description: OK
+          schema:
+            $ref: "#/definitions/Hub"
+        500:
+          $ref: '#/responses/InternalError'
+
```
Comment (lindenmckenzie) [inline swagger.yaml:83]: What's this for?

File: api/search_test.go [hunk 16]
```diff
@@ -387,6 +398,343 @@ func TestSearchHandlerFunc(t *testing.T) {
...
+		clList := ClientList{
+			scrubberClient: scrMock,
+			categoryClient: catMock,
+			berlinClient:   brMock,
+			dpESClient:     esMock,
+		}
+
+		cfg := &config.Config{
+			NlpToggle: true,
+			// NlpHubSettings: "{\"categoryWeighting\": 1000000000.0, \"categoryLimit\": 100, \"defaultState\": \"gb\"}",
```
Comment (lindenmckenzie) [inline api/search_test.go:518]: Why is this commented out?

File: api/search.go (hunk 5 above)
Comment (KamenDimitrov97) [reply api/search.go:161]: Because the golinter finds err to be a shadow variable. 
Do you want me to ignore the issue `//nolint:golint,unused` ?

Comment (lindenmckenzie) [reply api/search.go:161]: could just call it something more distinct like `validationErr`

File: swagger.yaml (hunk 15 above)
Comment (KamenDimitrov97) [reply swagger.yaml:83]: Testing Functionality to check if the NLP apis work and what responses they return, but we discussed it's removal I've forgotten to remove it from the documentation
Will do

File: README.md (hunk 1 above)
Comment (KamenDimitrov97) [reply README.md:42]: There are whitespace issues because the line itself is too long and the editors usually put it on the next line so that you can see it, but it's still on one line. If you open it in viewer mode or check it in the repo it's rendered correctly.
This is sort of out of my control 
Screenshot
